### PR TITLE
feat(fmgc): coarse lnav speed prediction

### PR DESCRIPTION
**Pull request by tracernz**
Fixes #[issue_no]

## Summary of Changes

Introduce a system to give coarse speed predictions for LNAV while we don't have proper predictions from VNAV. Use live wind speed to predict groundspeed for active/next legs.

Note: On top of #6804 .

## Screenshots (if necessary)

image

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Full flights needed. Check that turns predict reasonably, especially arrival/approach during cruise.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by TayDex** [comment]
What happened to this PR? While VNAV is implemented now, LNAV still doesn't use the speed predictions and this PR could help with the incorrect T/D placement due to currently more track miles being expected.

**Comment by tracernz** [comment]
> What happened to this PR? While VNAV is implemented now, LNAV still doesn't use the speed predictions and this PR could help with the incorrect T/D placement due to currently more track miles being expected.

People didn't like that SID turns were predicted with cruise speed when no constraints existed, so if I continue with this I will also need to do some coarse climb predictions.

**Comment by TayDex** [comment]
> People didn't like that SID turns were predicted with cruise speed when no constraints existed, so if I continue with this I will also need to do some coarse climb predictions.

Depending on how complex that it could be a temporary solution, else I guess we will have to wait for a proper implementation of the VNAV predictions in LNAV and wait with the T/D misplacement till then.

File: src/fmgc/src/guidance/vnav/AtmosphericConditions.ts [hunk 1]
```diff
@@ -22,7 +28,11 @@ export class AtmosphericConditions {
     update() {
         this.ambientTemperatureFromSim = SimVar.GetSimVarValue('AMBIENT TEMPERATURE', 'celsius');
         this.altitudeFromSim = SimVar.GetSimVarValue('INDICATED ALTITUDE', 'feet');
-        this.casFromSim = this.computeCasFromTas(this.altitudeFromSim, SimVar.GetSimVarValue('AIRSPEED TRUE', 'knots'));
+        this.tasFromSim = SimVar.GetSimVarValue('AIRSPEED TRUE', 'knots');
+        this.casFromSim = this.computeCasFromTas(this.altitudeFromSim, this.tasFromSim);
+        // TODO filter?
```
Comment (Saschl) [inline src/fmgc/src/guidance/vnav/AtmosphericConditions.ts:33]: Still relevant?

**Review by Saschl — APPROVED** [review]
Nice work. Cannot fully judge the changes logic wise, but code wise I'm happy 👍

File: src/fmgc/src/guidance/vnav/CoarsePredictions.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+            const alt = wp.additionalData.predictedAltitude;
+            const cas = wp.additionalData.predictedSpeed;
+            let tas = atmosphere.computeTasFromCas(alt, cas);
+            let gs = tas;
+
+            // predicted with live data for active and next two legs
+            if (i >= guidanceController.activeLegIndex && i < (guidanceController.activeLegIndex + 3)) {
+                tas = atmosphere.currentTrueAirspeed;
+                gs = tas + atmosphere.currentWindSpeed;
```
Comment (BlueberryKing) [inline src/fmgc/src/guidance/vnav/CoarsePredictions.ts:28]: Do you not have to take the course of the leg into account here? I believe this is what you have `getCurrentWindVelocityComponent` for.

**Review by BlueberryKing — COMMENTED** [review]
LGTM! I am wondering about one thing though. Does it make sense to use the currently measured wind to predict groundspeed in the arrival? The winds at cruise can be much higher and thus spoil your efforts in speed prediction using the speed constraints.

**Comment by tracernz** [comment]
> LGTM! I am wondering about one thing though. Does it make sense to use the currently measured wind to predict groundspeed in the arrival? The winds at cruise can be much higher and thus spoil your efforts in speed prediction using the speed constraints.

Current wind should only be used for next 3 legs, and for the rest GS = TAS... for now 😉 .

File: src/fmgc/src/guidance/vnav/AtmosphericConditions.ts (hunk 1 above)
Comment (tracernz) [reply src/fmgc/src/guidance/vnav/AtmosphericConditions.ts:33]: Yeah, although I think we should move all of the navigation stuff out to a navigation module that takes in data from ADIRUs etc. so I won't do it now.

**Comment by John-UK** [comment]
Quality Assurance Report
Discord john P Maguire#9027
Object of testing: #6808
Tier of Testing: 2
Date: 05/04/2022

Testing Process:

Flew the route in experimental build a70553c6. Noted the ND displayed routing errors related to using cruise speeds for turns. Picture 1 DLs.
unknown

Installed PR #6808 and spawned C&D at EGJJ stand 7. Completed a full procedure preparation using the same parameters as used in experimental flight. Picture 2 at the stand.
egll1
The 999nm between waypoints was not seen in this build. 

A small break in the nav track is seen at OCK07. This small curve had no effect on the AP and the approach was flown as expected. in pic 3 small curve on nav track at OCK. 
egll4

Negatives
Not directly related. The departure, climb, cruise and descent to OCK was similar to experimental. There were no top of climb or top of descent symbols. 

Testing Results
Passed. 

Conclusions
LGTM. The use of a lower predicted speed for the approach turns corrected the nav track from OCK to the ILS. .  Not sure on what caused the small curve at OCK all the distances in the MCDU were correct with the chart.

**Comment by John-UK** [comment]
Quality Assurance Report
Discord john P Maguire#9027
Object of testing: #6808
Tier of Testing: 2
Date: 05/04/2022

Testing Process:

Flew the route in experimental build a70553c6. Noted the ND displayed routing errors related to using cruise speeds for turns. Picture 1 DLs.
unknown

Installed PR #6808 and spawned C&D at EGJJ stand 7. Completed a full procedure preparation using the same parameters as used in experimental flight. Picture 2 at the stand.
egll1
The 999nm between waypoints was not seen in this build. 

A small break in the nav track is seen at OCK07. This small curve had no effect on the AP and the approach was flown as expected. in pic 3 small curve on nav track at OCK. 
egll4

Negatives
Not directly related. The departure, climb, cruise and descent to OCK was similar to experimental. There were no top of climb or top of descent symbols. 
Testing Results
Passed. 
Conclusions
The use of a lower predicted speed for the approach turns corrected the nav track from OCK to the ILS. LGTM.  Not sure on what caused the small curve at OCK all the distances in the MCDU were correct with the chart.

**Comment by mico975** [comment]
Quality Assurance Tester Report

Discord          : mico#3145
Object of testing: #6808
Tier of Testing  : 2
Date             : 06/04/2022

Testing Process: 
Programmed various STARs for multiple airports while in cruise. Turn predictions on STARs are much more reasonable now.
Full flight with landing at EGKK went well.

Negatives:       
N/A

Testing Results: 
Passed

Conclusions: 
LGTM

Media:       
image

**Pull request merged**